cloud/openstack/os_floating_ip.py

```diff
@@ -0,0 +1,245 @@
+- os_floating_ip:
+     cloud: mordred
+     state: present
+     server: vm1
+     network_name: external_network
+     internal_network_name: internal_network
+'''
+
+
+def _get_server_state(module, cloud):
+    info = None
+    server = cloud.get_server(module.params['server'])
+    if server:
+        info = server._info
+        status = info['status']
+        if status != 'ACTIVE' and module.params['state'] == 'present':
+            module.fail_json(
+                msg="The VM is available but not Active. State: %s" % status
+            )
+    return info, server
+
+
+def _get_port_info(neutron, module, instance_id, internal_network_name=None):
+    subnet_id = None
+    if internal_network_name:
+        kwargs = {'name': internal_network_name}
+        networks = neutron.list_networks(**kwargs)
+        network_id = networks['networks'][0]['id']
+        kwargs = {
+            'network_id': network_id,
+            'ip_version': 4
+        }
+        subnets = neutron.list_subnets(**kwargs)
+        subnet_id = subnets['subnets'][0]['id']
+
+    kwargs = {
+        'device_id': instance_id,
+    }
+    try:
+        ports = neutron.list_ports(**kwargs)
+    except Exception, e:
+        module.fail_json(msg="Error in listing ports: %s" % e.message)
+
+    if subnet_id:
+        port = next(port for port in ports['ports'] if port['fixed_ips'][0]['subnet_id'] == subnet_id)
+        port_id = port['id']
+        fixed_ip_address = port['fixed_ips'][0]['ip_address']
+    else:
+        port_id = ports['ports'][0]['id']
+        fixed_ip_address = ports['ports'][0]['fixed_ips'][0]['ip_address']
+
+    if not ports['ports']:
+        return None, None
+    return fixed_ip_address, port_id
+
+
+def _get_floating_ip(neutron, module, fixed_ip_address):
+    kwargs = {
+        'fixed_ip_address': fixed_ip_address
+    }
+    try:
+        ips = neutron.list_floatingips(**kwargs)
+    except Exception, e:
+        module.fail_json(
+            msg="Error in fetching the floatingips's %s" % e.message
+        )
+
+    if not ips['floatingips']:
+        return None, None
+
+    return (ips['floatingips'][0]['id'],
+            ips['floatingips'][0]['floating_ip_address'])
+
+
+def _create_and_associate_floating_ip(neutron, module, port_id,
+                                      net_id, fixed_ip):
+    kwargs = {
+        'port_id': port_id,
+        'floating_network_id': net_id,
+        'fixed_ip_address': fixed_ip
+    }
+
+    try:
+        result = neutron.create_floatingip({'floatingip': kwargs})
+    except Exception, e:
+        module.fail_json(
+            msg="Error in updating the floating ip address: %s" % e.message
+        )
+
+    module.exit_json(
+        changed=True,
+        result=result,
+        public_ip=result['floatingip']['floating_ip_address']
+    )
+
+
+def _get_public_net_id(neutron, module):
+    kwargs = {
+        'name': module.params['network_name'],
+    }
```

Comment:
Is there a way to take advantage of the default_floating_pool info that exists within nova? I think if a source pool isn't defined by the client, the server will attempt to use default_floating_pool, but if it's not configured you'll get an error. Perhaps we can take advantage of that here? Just a thought.